File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        vector<vector<int>> dp(n,vector<int> (m,0));
        dp[0][0]=grid[0][0];
        for(int i=1;i<m;i++){
            dp[0][i]=grid[0][i]+dp[0][i-1];
        }
         for(int i=1;i<n;i++){
            dp[i][0]=grid[i][0]+dp[i-1][0];
        }
        for(int j=1;j<n;j++){
            for(int k=1;k<m;k++){
                dp[j][k]=grid[j][k]+min(dp[j][k-1],dp[j-1][k]);
            }
        }
        return dp[n-1][m-1];
    }
};

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int[][] dp=new int[n][m];
        for(int i=0;i<n;i++){
            for(int  j=0;j<m;j++){
                if(i==0 &&j==0) dp[0][0]=grid[0][0];
                else{
                    int up=Integer.MAX_VALUE,left=Integer.MAX_VALUE;
                if(i>0) up=dp[i-1][j];
                if(j>0) left=dp[i][j-1];
                dp[i][j]=Math.min(left,up)+grid[i][j];
                }
            }
        }
        return dp[n-1][m-1];
    }
}